kotlin-web-spring-boot-3/src/main/kotlin/com/vndevteam/kotlinwebspringboot3/application/logging/RequestFilter.kt

```diff
@@ -0,0 +1,104 @@
+import org.slf4j.MDC
+import org.springframework.web.filter.OncePerRequestFilter
+import java.util.*
+
+class RequestFilter : OncePerRequestFilter {
+
+    companion object {
+        const val REQUEST_ID = "request-id"
+        const val START_TIME = "start-time"
+        private const val HEADER_NAME_API_KEY = "apikey"
+        private const val X_FORWARDED_FOR = "X-Forwarded-For"
+        val log: Logger = LoggerFactory.getLogger(RequestFilter::class.java)
+    }
+
+    private var responseHeader: String? = null
+    private var mdcTokenKey: String? = null
+    private var mdcClientIpKey: String? = null
+    private var requestHeader: String? = null
+
+    constructor() {
+        this.responseHeader = Slf4jMDCFilterConfiguration.DEFAULT_RESPONSE_TOKEN_HEADER
+        this.mdcTokenKey = Slf4jMDCFilterConfiguration.DEFAULT_MDC_UUID_TOKEN_KEY
+        this.mdcClientIpKey = Slf4jMDCFilterConfiguration.DEFAULT_MDC_CLIENT_IP_KEY
+        this.requestHeader = null
+    }
+
+    constructor(responseHeader: String, mdcTokenKey: String, mdcClientIpKey: String, requestHeader: String?) {
+        this.responseHeader = responseHeader
+        this.mdcTokenKey = mdcTokenKey
+        this.mdcClientIpKey = mdcClientIpKey
+        this.requestHeader = requestHeader
+    }
+
+    override fun doFilterInternal(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        filterChain: FilterChain
+    ) {
+        try {
+            val clientId: String = getClientId(request)
+            val clientIp: String = getClientIpAddress(request)
+            MDC.put(mdcClientIpKey, clientIp)
+            MDC.put(mdcTokenKey, clientId)
+
+            if (!responseHeader.isNullOrBlank()) {
+                response.addHeader(responseHeader, clientId)
+            }
+
+            request.setAttribute(REQUEST_ID, clientId)
+            request.setAttribute(START_TIME, System.currentTimeMillis())
+            logRequest(request, clientId)
+            filterChain.doFilter(request, response)
+        } finally {
+            MDC.remove(mdcTokenKey)
+            MDC.remove(mdcClientIpKey)
+        }
+    }
+
+    private fun logRequest(request: HttpServletRequest?, requestId: String) {
+        if (request != null) {
+            val data = StringBuilder()
+                .append("[PATH]: ").append(request.requestURI).append(". ")
+                .append("[QUERIES]: ").append(request.queryString).append(". ")
+                .append("[HEADERS]: ")
+
+            val headerNames = request.headerNames
+            while (headerNames.hasMoreElements()) {
+                val key = headerNames.nextElement() as String
+                // Do not log sensitive header: such as apikey
+                if (key != HEADER_NAME_API_KEY) {
+                    val value = request.getHeader(key)
+                    data.append(" --- ").append(key).append(" : ").append(value)
+                }
+            }
+
+            log.info(data.toString())
```

Comment:
@haiks-2382
```suggestion
            log.debug(data.toString())
```
Chỉ nên bật khi debug

Reply:
t dùng config tùy lúc truyền vào application.yml để quyết định có log ra ko ô xem ok k nhé